000-mm2-connector-offsets-support.md

```diff
@@ -0,0 +1,390 @@
+
+### Request/Response from the API endpoints
+
+The response body received when using the GET endpoint, and the request body used when calling the PATCH endpoint have identical structures to allow users to easily fetch and then alter offsets.
+The structure of the body is not fully defined for generic source connectors.
+This is because source connectors determine their own structure for offsets depending on the source system they are integrating with.
+However, for each of the MirrorMaker connectors (MirrorSourceConnector, MirrorCheckpointConnector, MirrorHeartbeatConnector) we know the expected structure.
+
+MirrorSourceConnector format:
+```
+{
+  "offsets": [
+    {
+      "partition": {
+        "cluster": "east-kafka",
+        "partition": 0,
+        "topic": "mirrormaker2-cluster-configs"
+      },
+      "offset": {
+        "offset": 0
+      }
+    }
+  ]
+}
+```
+
+MirrorCheckpointConnector format:
+```
+{
+  "offsets": [
+    {
+      "partition": {
+        "partition": 4,
+        "topic": "inventory",
+        "group": "my-group"
+      },
+      "offset": {
+        "offset": 0
+      }
+    }
+  ]
+}
+```
+
+MirrorHeartbeatConnector format:
+```
+{
+  "offsets": [
+    {
+      "partition": {
+        "sourceClusterAlias": "east-kafka",
+        "targetClusterAlias": "west-kafka"
+      },
+      "offset": {
+        "offset": 0
+      }
+    }
+  ]
+}
+```
+
+To alter offsets we do not need the user to supply the entirety of these JSON objects.
+That is because some of the fields are already known to the operator.
+Specifically, we can omit:
+* For MirrorSourceConnector:
+  * `cluster`, this is equal to `spec.mirrors[].sourceCluster` in the KafkaMirrorMaker2 resource
+  * `topic`, this is the configuration topic for the underlying Kafka Connect cluster, by default named `mirrormaker2-cluster-configs`
+* For MirrorCheckpointConnector no fields can be omitted
+* For MirrorHeartbeatConnector:
+  * `partition.sourceClusterAlias`, this is equal to `spec.mirrors[].sourceCluster` in the KafkaMirrorMaker2 resource
+  * `partition.targetClusterAlias`, this is equal to `spec.mirrors[].targetCluster` in the KafkaMirrorMaker2 resource
+
+We will also omit these from the output when listing offsets.
+This will make it easier for users to reuse the list offsets output when altering offsets.
+
+### New annotations
+
+All three actions will make use of a new annotation called `strimzi.io/connector-offsets`.
+The possible values will be `list`, `alter`, and `reset`.
+
+A second new annotation will be added called `strimzi.io/mirrormaker-connector`.
+It will be required when `strimzi.io/connector-offsets` is set.
+The value will be the name of the connector to apply the action to, for example `east-kafka->west-kafka.MirrorSourceConnector`.
+
+If the `strimzi.io/mirrormaker-connector` annotation is set, but the `strimzi.io/connector-offsets` annotation is missing the operator will take no action.
```

Comment:
Isn't it better to have a single annotation with some JSON structure? What if you would want to e.g. reset source and checkpoint connectors at the same time which seem slike something what might make sense?

Reply:
My understanding of Kubernetes annotations is they are intended to be a key/value pair where the key and value are both strings, are you suggesting the value string actually be JSON, e.g.
```
strimzi.io/mm2-connector-offsets: "{[{\"connector\": \"east-kafka->west-kafka.MirrorSourceConnector\", \"action\": \"list\"}, {\"connector\": \"east-kafka->west-kafka.MirrorCheckpointConnector\", \"action\": \"list\"}]}"
```

I worry that makes it quite a lot of work for the user to construct the JSON and a lot of room for error.

I did wonder about allowing the `strimzi.io/mirrormaker-connector` to be a comma separated list, but it brings the same difficulty that I listed in the rejected alternatives around making the `mirrormaker-connector` annotation optional, that the reconciliation is performed per connector and so it requires error handling if the action on one connector succeeds and one doesn't.

However, if we think it's a likely usecase to be able to apply actions to multiple connectors at once I can describe that error handling, I think it would likely just mean if something goes wrong the user has to reannotate the resource, so it's up to them to check the status for what succeeded and what didn't.

Reply:
I don't know, to me, this seems like something what make sense?
* When using multiple source clusters, I might want to reset all of their source connector offsets at once?
* When using source and offset mirroring, I might want to reset both the source and checkpoint connector offsets at the same point?

The point about JSON in the annotation is partially valid. Storing JSONs in annotation is common. But it does make it hard to annotate it from the command line (but is that how users will annotate it? Is it a bigger problem than having two annotations to be set, which if they are set one by one also causes problems and errors?).

Reply:
I'm happy to update the proposal so multiple connectors can be actioned at once, but I'd personally advocate for keeping with the two annotations, rather than having a more complex single annotation. I don't think there are big problems with if one is present and one not:

| `strimzi.io/connector-offsets` | `strimzi.io/mirrormaker-connectors` | outcome |
| -- | -- | -- |
| present | present | action applied |
| missing | missing | nothing happens |
| missing | present | nothing happens |
| present | missing | ? |

It's only the last state which is potentially a problem. For that I think either we could add a warning as I described, or if we are supporting multiple connectors, this could apply the action to all connectors in all mirrors.

You can use the `kubectl annotate` command to add multiple annotations at once

Reply:
I don't think applying it to all connectors makes sense. That would be a big risk. Especially if you expect both annotation to be used. I would also expect at least a warning in the other case when only one annotation is present.